src/jdk.httpserver/share/classes/sun/net/httpserver/ServerImpl.java

```diff
@@ -965,17 +965,20 @@ void requestStarted (HttpConnection c) {
     void markIdle(HttpConnection c) {
         Boolean close = false;
 
-        idleConnectionLock.lock();
-        if (idleConnections.size() >= MAX_IDLE_CONNECTIONS) {
-            // closing the connection here could block
-            // instead set boolean and close outside of lock
-            close = true;
-        } else {
-            c.idleStartTime = System.currentTimeMillis();
-            c.setState(State.IDLE);
-            idleConnections.add(c);
+        try {
+            idleConnectionLock.lock();
```

Comment:
The lock shoud be outside of the try { } finally { } block: if lock() fails, then the lock hasn't been acquired and therefore you shouldn't try to release it.